src/game/BattleGround/BattleGroundWS.cpp

```diff
@@ -217,218 +293,143 @@ void BattleGroundWS::EventPlayerCapturedFlag(Player* source)
-            && event == WS_EVENT_FLAG_A)
+    // check if we are returning our flag
+    if (wsFlagIds[teamIdx] == displayId)
     {
-        message_id = LANG_BG_WS_PICKEDUP_AF;
-        type = CHAT_MSG_BG_SYSTEM_HORDE;
-        PlaySoundToAll(BG_WS_SOUND_ALLIANCE_FLAG_PICKED_UP);
-        SpawnEvent(WS_EVENT_FLAG_A, 0, false);
-        SetAllianceFlagCarrier(source->GetObjectGuid());
-        m_FlagState[TEAM_INDEX_ALLIANCE] = BG_WS_FLAG_STATE_ON_PLAYER;
-        // update world state to show correct flag carrier
-        UpdateFlagState(HORDE, BG_WS_FLAG_STATE_ON_PLAYER);
-        UpdateWorldState(BG_WS_FLAG_UNK_ALLIANCE, 1);
-        source->CastSpell(source, BG_WS_SPELL_SILVERWING_FLAG, TRIGGERED_OLD_TRIGGERED);
+        actionId = BG_WS_FLAG_ACTION_RETURNED;
+        Team team = player->GetTeam();
+        UpdateFlagState(GetOtherTeam(team), BG_WS_FLAG_STATE_WAIT_RESPAWN);
+        RespawnFlag(team, false);
+        UpdatePlayerScore(player, SCORE_FLAG_RETURNS, 1);
+        SendMessageToAll(wsMessageIds[teamIdx][actionId], wsChatMessageTypes[teamIdx][actionId], player);
+        PlaySoundToAll(wsSounds[teamIdx][actionId]);
     }
-
-    // horde flag picked up from base
-    if (source->GetTeam() == ALLIANCE && GetFlagState(HORDE) == BG_WS_FLAG_STATE_ON_BASE
-            && event == WS_EVENT_FLAG_H)
+    // check if we are picking up enemy flag
+    else if (wsFlagIds[otherTeamIdx] == displayId)
     {
-        message_id = LANG_BG_WS_PICKEDUP_HF;
-        type = CHAT_MSG_BG_SYSTEM_ALLIANCE;
-        PlaySoundToAll(BG_WS_SOUND_HORDE_FLAG_PICKED_UP);
-        SpawnEvent(WS_EVENT_FLAG_H, 0, false);
-        SetHordeFlagCarrier(source->GetObjectGuid());
-        m_FlagState[TEAM_INDEX_HORDE] = BG_WS_FLAG_STATE_ON_PLAYER;
-        // update world state to show correct flag carrier
-        UpdateFlagState(ALLIANCE, BG_WS_FLAG_STATE_ON_PLAYER);
-        UpdateWorldState(BG_WS_FLAG_UNK_HORDE, 1);
-        source->CastSpell(source, BG_WS_SPELL_WARSONG_FLAG, TRIGGERED_OLD_TRIGGERED);
+        actionId = BG_WS_FLAG_ACTION_PICKEDUP;
+        SpawnEvent(otherTeamIdx, 0, false);
+        SetFlagCarrier(otherTeamIdx, player->GetObjectGuid());
+        player->CastSpell(player, wsSpellTypes[otherTeamIdx][BG_WS_FLAG_ACTION_PICKEDUP], TRIGGERED_OLD_TRIGGERED);
+        m_FlagState[otherTeamIdx] = BG_WS_FLAG_STATE_ON_PLAYER;
+        UpdateFlagState(player->GetTeam(), BG_WS_FLAG_STATE_ON_PLAYER);
+        UpdateWorldState(wsStateUpdateId[otherTeamIdx], 1);
+        SendMessageToAll(wsMessageIds[otherTeamIdx][actionId], wsChatMessageTypes[teamIdx][actionId], player);
+        PlaySoundToAll(wsSounds[otherTeamIdx][actionId]);
     }
 
-    // Alliance flag on ground(not in base) (returned or picked up again from ground!)
-    if (GetFlagState(ALLIANCE) == BG_WS_FLAG_STATE_ON_GROUND && source->IsWithinDistInMap(target_obj, 10))
+    if (actionId > -1)
+        player->RemoveAurasWithInterruptFlags(AURA_INTERRUPT_FLAG_ENTER_PVP_COMBAT);
```

Comment:
You could still remove this branch by moving this into each section of the previous branch which is more efficient and easier to understand

Reply:
I think since I'm returning what action was executed, I might as well leave this as it is.